### PR TITLE
Improve robustness of LocalCacheManager

### PR DESCRIPTION
**Pull request by apc999**
This PR fixes a potential race condition in LocalCacheManager and improves its robustness.

Though designed to be threadsafe, `LocalCacheManager` returns a `ReadableByteChannel` to 
`LocalCacheFileInStream`, where I/O operations like reading a file is actually done at `read` or `positionedRead` at `LocalCacheFileInStream`. This leaves an opportunity to multiple threads racing on the same `file` outside threadsafe `LocalCacheManager`. E.g., a super slow reader thread may contend with a evictor thread who wants to remove the same page file per LRU policy, leaving the reader thread risk undefined behavior (up to operating system).

Invariants after this PR:

(1) operating a page files is always protected inside the page lock.
(2) given no operation on the cache `LocalCacheFileInStream` (moved to `LocalCacheManager` in this PR), if cache fails in certain unexpected way, exceptions will always be catched by the `NoExceptionCacheManager`.

This PR changes the API of `LocalCacheManager.read` from returning a channel to returning the number of bytes read but copying bytes inside, which is a bit different from what i previously discussed with @calvinjia . This will make the code much simpler.

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 1]
```diff
@@ -132,47 +134,40 @@ public int read(byte[] b, int off, int len) throws IOException {
...
     // for each page, check if it is available in the cache
-    while (bytesRead < lengthToRead) {
+    while (totalBytesRead < lengthToRead) {
       long currentPage = mPosition / mPageSize;
       int currentPageOffset = (int) (mPosition % mPageSize);
-      int bytesLeftInPage = (int) Math.min(mPageSize - currentPageOffset, lengthToRead - bytesRead);
+      int bytesLeftInPage =
+          (int) Math.min(mPageSize - currentPageOffset, lengthToRead - totalBytesRead);
       PageId pageId = new PageId(mStatus.getFileIdentifier(), currentPage);
-      try (ReadableByteChannel cachedData = mCacheManager.get(pageId, currentPageOffset)) {
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:143]: move the IO operations inside `LocalCacheManager` under page locks to coordinate with other threads, also procted by `NoExceptionManager`

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java [hunk 2]
```diff
@@ -445,18 +445,33 @@ private boolean deletePage(PageId pageId, PageInfo pageInfo) {
...
+        Metrics.GET_ERRORS_FAILED_READ.inc();
+        throw new IOException(String.format(
+            "Failed to read page {}: supposed to read {} bytes, {} bytes actually read",
+            pageId, bytesToRead, buf.position() - offsetInBuffer));
+      }
     } catch (IOException | PageNotFoundException e) {
       LOG.error("Failed to get existing page {}: {}", pageId, e);
       Metrics.GET_ERRORS.inc();
-      return null;
+      return -1;
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:471]: Should we have this logic here or leave it to `NoExceptionCacheManager`?

Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:471]: it needs to be here so we can clean up the state https://github.com/Alluxio/alluxio/pull/11525/files#diff-52b3926ca60dbb1ab7f30f23ada07efdR365

Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:471]: got it, thanks.

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by jainxrohit** [comment]
@apc999  Can we merge this to branch-2.2 as well?

**Comment by apc999** [comment]
alluxio-bot, cherry-pick this to branch-2.2 please

**Pull request merged**